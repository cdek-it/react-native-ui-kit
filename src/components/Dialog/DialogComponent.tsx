import React, { type ReactNode } from 'react'
import { View, useWindowDimensions } from 'react-native'

import { makeStyles } from '../../utils/makeStyles'

export interface DialogComponentProps {
  readonly body?: React.ComponentType
  readonly footer?: React.ComponentType
  readonly header?: ReactNode
}

export const DialogComponent: React.FC<DialogComponentProps> = ({
  body: Body,
  footer: Footer,
  header = null,
}) => {
  const styles = useStyles()
  const { width, height } = useWindowDimensions()
  const maxSize = { maxWidth: width - 40, maxHeight: height - 100 }

  return (
    <View style={[styles.root, maxSize]}>
      {header}

      {Body ? (
        <View style={styles.body}>
          <Body />
        </View>
      ) : null}

      {Footer ? (
        <View style={styles.footer}>
          <Footer />
        </View>
      ) : null}
    </View>
  )
}

const useStyles = makeStyles(({ theme }) => ({
  root: {
    backgroundColor: theme.Overlay.Dialog.Header.dialogHeaderBg,
    borderColor: theme.Overlay.Overlay.overlayContentBorderColor,
    borderRadius: theme.General.borderRadius,
  },
  body: {
    paddingVertical: theme.Overlay.Dialog.Body.dialogContentPaddingTopBottom,
    paddingHorizontal: theme.Overlay.Dialog.Body.dialogContentPaddingLeftRight,
  },
  footer: {
    paddingBottom: theme.Overlay.Dialog.Footer.dialogFooterPaddingLeftRight,
    paddingHorizontal: theme.Overlay.Dialog.Footer.dialogFooterPaddingLeftRight,
  },
}))
