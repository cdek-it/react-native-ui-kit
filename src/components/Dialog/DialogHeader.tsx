import {
  type Icon,
  IconAlertTriangle,
  IconCircleCheck,
  IconCircleX,
  IconHelpCircle,
  IconInfoCircle,
  IconX,
} from '@tabler/icons-react-native'
import { useMemo } from 'react'
import { TouchableOpacity, View } from 'react-native'

import { makeStyles } from '../../utils/makeStyles'
import { Title } from '../Typography'

type Severity = 'danger' | 'warning' | 'info' | 'success' | 'help'

export interface DialogHeaderProps {
  readonly title: string
  readonly onClose?: () => void
  readonly severity?: Severity
}

const iconsMap: Record<Severity, Icon> = {
  danger: IconCircleX,
  warning: IconAlertTriangle,
  info: IconInfoCircle,
  success: IconCircleCheck,
  help: IconHelpCircle,
}

export const DialogHeader = ({
  title,
  onClose,
  severity,
}: DialogHeaderProps) => {
  const styles = useStyles()
  const tids = DialogHeaderTestId

  const icon = useMemo(() => {
    if (!severity) return null

    const Icon = iconsMap[severity]

    return <Icon {...styles.severityIcon} {...styles[severity]} />
  }, [severity, styles])

  return (
    <>
      <View style={styles.header}>
        {icon}
        <Title level='h3' style={styles.text} testID={tids.title}>
          {title}
        </Title>
        {onClose ? (
          <TouchableOpacity
            style={styles.closeTouchable}
            testID={tids.closeButton}
            onPress={onClose}
          >
            <IconX {...styles.closeIcon} />
          </TouchableOpacity>
        ) : null}
      </View>

      <View style={styles.separator} />
    </>
  )
}

const useStyles = makeStyles(({ theme, spacing, typography }) => ({
  text: { flex: 1 },
  header: {
    flexDirection: 'row',
    gap: spacing.Gap['gap-4'],
    alignItems: 'center',
    paddingBottom: theme.Overlay.Dialog.Header.dialogHeaderPaddingTopBottom,
    padding: theme.Overlay.Dialog.Header.dialogHeaderPaddingLeftRight,
  },
  closeTouchable: { padding: 8 },
  separator: {
    height: 1,
    backgroundColor: theme.Overlay.Overlay.overlayContentBorderColor,
  },
  closeIcon: { width: 14, height: 14, color: theme.General.actionIconColor },
  severityIcon: {
    width: typography.Size['text-2xl'],
    height: typography.Size['text-2xl'],
  },
  danger: { color: typography.Color.Service['text-danger'] },
  warning: { color: typography.Color.Service['text-warning'] },
  info: { color: typography.Color.Service['text-info'] },
  success: { color: typography.Color.Service['text-success'] },
  help: { color: typography.Color.Service['text-help'] },
}))

export const DialogHeaderTestId = {
  title: 'DialogTitle',
  closeButton: 'DialogCloseButton',
}
