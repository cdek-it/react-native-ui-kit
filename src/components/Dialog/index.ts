export { Dialog, type DialogProps, DialogTestId } from './Dialog'
export { DialogComponent, type DialogComponentProps } from './DialogComponent'
export {
  DialogHeader,
  type DialogHeaderProps,
  DialogHeaderTestId,
} from './DialogHeader'
