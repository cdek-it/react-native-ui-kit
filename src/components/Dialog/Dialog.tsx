import { Portal } from '@gorhom/portal'
import React, { useCallback, useEffect } from 'react'
import { View, StyleSheet, BackHandler, Pressable } from 'react-native'
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  withSpring,
  runOnJS,
  interpolate,
} from 'react-native-reanimated'

import { DialogComponent, type DialogComponentProps } from './DialogComponent'

const ANIMATION_DURATION = 200
const BACKDROP_OPACITY = 0.5
const SCALE_DAMPING = 5
const SCALE_STIFFNESS = 50
const SCALE_INIT_VALUE = 0.9

export interface DialogProps extends DialogComponentProps {
  readonly isVisible: boolean
  readonly onClose?: () => void
  readonly onHideComplete?: () => void
  readonly testID?: string
}

// eslint-disable-next-line import-x/no-deprecated
const AnimatedPressable = Animated.createAnimatedComponent(Pressable)

export const Dialog: React.FC<DialogProps> = ({
  isVisible,
  onClose,
  onHideComplete,
  header,
  footer,
  body,
  testID,
}) => {
  const opacity = useSharedValue(0)
  const scale = useSharedValue(SCALE_INIT_VALUE)

  const handleAnimationComplete = useCallback(() => {
    if (onHideComplete) {
      onHideComplete()
    }
  }, [onHideComplete])

  useEffect(() => {
    if (isVisible) {
      opacity.value = withTiming(1, { duration: ANIMATION_DURATION })
      scale.value = withSpring(1, {
        damping: SCALE_DAMPING,
        stiffness: SCALE_STIFFNESS,
      })
    } else {
      opacity.value = withTiming(0, { duration: ANIMATION_DURATION })
      scale.value = withTiming(
        SCALE_INIT_VALUE,
        { duration: ANIMATION_DURATION },
        (finished) => {
          if (finished) {
            runOnJS(handleAnimationComplete)()
          }
        }
      )
    }
  }, [isVisible, opacity, scale, handleAnimationComplete])

  const backdropAnimatedStyle = useAnimatedStyle(() => {
    return {
      opacity: interpolate(opacity.value, [0, 1], [0, BACKDROP_OPACITY]),
    }
  })

  const dialogAnimatedStyle = useAnimatedStyle(() => {
    return { opacity: opacity.value, transform: [{ scale: scale.value }] }
  })

  useEffect(() => {
    const subscription = BackHandler.addEventListener(
      'hardwareBackPress',
      () => {
        if (isVisible) {
          onClose?.()

          return true
        }

        return false
      }
    )

    return subscription.remove
  }, [isVisible, onClose])

  return (
    <Portal>
      <View
        style={[
          styles.container,
          !isVisible && styles.containerNoPointerEvents,
        ]}
        testID={testID ?? DialogTestId.root}
      >
        <AnimatedPressable
          style={[styles.backdrop, backdropAnimatedStyle]}
          testID={DialogTestId.backdrop}
          onPress={onClose}
        />
        <Animated.View
          style={dialogAnimatedStyle}
          testID={DialogTestId.contentContainer}
        >
          <DialogComponent body={body} footer={footer} header={header} />
        </Animated.View>
      </View>
    </Portal>
  )
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    zIndex: 9999,
    elevation: 999,
    pointerEvents: 'auto',
  },
  containerNoPointerEvents: { pointerEvents: 'none' },
  backdrop: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    backgroundColor: 'black',
  },
})

export const DialogTestId = {
  root: 'DialogModal',
  backdrop: 'Backdrop',
  contentContainer: 'ContentContainer',
}
