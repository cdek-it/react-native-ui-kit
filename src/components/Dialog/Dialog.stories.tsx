import type { Meta, StoryObj } from '@storybook/react'
import { useCallback, useState } from 'react'
import { Text, TouchableOpacity, View } from 'react-native'

import { Dialog, type DialogProps } from './Dialog'
import { DialogHeader, type DialogHeaderProps } from './DialogHeader'

const severityMap = {
  info: 'info',
  success: 'success',
  warning: 'warning',
  danger: 'danger',
  help: 'help',
  none: undefined,
}

const meta: Meta<DialogProps & DialogHeaderProps> = {
  title: 'Dialog',
  component: Dialog,
  render: (props) => Template(props),
  argTypes: {
    severity: {
      control: 'radio',
      options: Object.keys(severityMap),
      mapping: severityMap,
    },
    onClose: { action: 'onClose' },
    onHideComplete: { action: 'onHideComplete' },
  },
}

const Template = ({
  severity,
  onHideComplete,
  onClose,
}: DialogProps & DialogHeaderProps) => {
  const [isVisible, setIsVisible] = useState(false)

  const showDialog = () => {
    setIsVisible(true)
  }

  const hideDialog = useCallback(() => {
    setIsVisible(false)
    onClose?.()
  }, [onClose])

  const Body = useCallback(() => {
    return (
      <Text>Войдите в приложение, чтобы заказ сохранился в списке заказов</Text>
    )
  }, [])

  const Footer = useCallback(() => {
    return (
      <TouchableOpacity onPress={hideDialog}>
        <Text>Войти</Text>
      </TouchableOpacity>
    )
  }, [hideDialog])

  return (
    <View>
      <View>
        <TouchableOpacity onPress={showDialog}>
          <Text>Показать Dialog</Text>
        </TouchableOpacity>
      </View>

      <Dialog
        body={Body}
        footer={Footer}
        header={
          <DialogHeader
            severity={severity}
            title='Найденный заказ не сохранится'
            onClose={hideDialog}
          />
        }
        isVisible={isVisible}
        onClose={hideDialog}
        onHideComplete={onHideComplete}
      />
    </View>
  )
}

export default meta

type Story = StoryObj<typeof Dialog>

const DialogStory: Story = { args: {} }

export { DialogStory as Dialog }
