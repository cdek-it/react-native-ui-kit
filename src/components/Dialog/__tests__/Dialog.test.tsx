import { PortalProvider } from '@gorhom/portal'
import { render } from '@testing-library/react-native'

import { Text } from 'react-native'

import { Dialog, type DialogProps } from '../Dialog'

describe('Dialog component tests', () => {
  const MockBody = () => <Text>This is the dialog body content</Text>
  const MockFooter = () => <Text>This is the dialog footer</Text>
  const header = <Text>This is custom dialog header</Text>

  const snapshotCases: Array<[string, DialogProps]> = [
    [
      'basic dialog with title, body and footer',
      {
        isVisible: true,
        header,
        body: MockBody,
        footer: MockFooter,
        onClose: jest.fn(),
      },
    ],
  ]

  test.each(snapshotCases)('%s', (_, props) => {
    const renderedDialog = render(
      <PortalProvider>
        <Dialog {...props} />
      </PortalProvider>
    )

    expect(renderedDialog.toJSON()).toMatchSnapshot()
  })
})
