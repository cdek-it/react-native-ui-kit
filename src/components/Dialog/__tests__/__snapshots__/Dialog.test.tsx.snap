// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog component tests basic dialog with title, body and footer 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "bottom": 0,
        "elevation": 999,
        "flex": 1,
        "justifyContent": "center",
        "left": 0,
        "pointerEvents": "auto",
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 9999,
      },
      false,
    ]
  }
  testID="DialogModal"
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    jestAnimatedStyle={
      {
        "value": {
          "opacity": 0,
        },
      }
    }
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "black",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="Backdrop"
  />
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "opacity": 0,
          "transform": [
            {
              "scale": 0.9,
            },
          ],
        },
      }
    }
    style={
      {
        "opacity": 0,
        "transform": [
          {
            "scale": 0.9,
          },
        ],
      }
    }
    testID="ContentContainer"
  >
    <View
      style={
        [
          {
            "backgroundColor": "#ffffff",
            "borderColor": "#dbdada",
            "borderRadius": 7,
          },
          {
            "maxHeight": 1234,
            "maxWidth": 710,
          },
        ]
      }
    >
      <Text>
        This is custom dialog header
      </Text>
      <View
        style={
          {
            "paddingHorizontal": 21,
            "paddingVertical": 21,
          }
        }
      >
        <Text>
          This is the dialog body content
        </Text>
      </View>
      <View
        style={
          {
            "paddingBottom": 21,
            "paddingHorizontal": 21,
          }
        }
      >
        <Text>
          This is the dialog footer
        </Text>
      </View>
    </View>
  </View>
</View>
`;
